### PR TITLE
Don't call malloc from freeze which is called in a signal handler

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

File: src/shared/mount-util.c [hunk 1]
```diff
@@ -1010,11 +1011,9 @@ static int make_userns(uid_t uid_shift, uid_t uid_range) {
         r = safe_fork("(sd-mkuserns)", FORK_CLOSE_ALL_FDS|FORK_DEATHSIG|FORK_NEW_USERNS, &pid);
         if (r < 0)
                 return r;
-        if (r == 0) {
+        if (r == 0)
                 /* Child. We do nothing here, just freeze until somebody kills us. */
                 freeze();
-                _exit(EXIT_FAILURE);
```
Comment (bluca) [inline src/shared/mount-util.c:1016]: wouldn't it be better to leave this? Just in case - doesn't really hurt, no?

**Review by bluca — APPROVED** [review]
Changes look good to me. If close_range is available I don't think it's currently possible to malloc from the signal handler, but given the fallback seems simple enough it's nice to have for older kernels.

File: src/shared/mount-util.c (hunk 1 above)
Comment (keszybz) [reply src/shared/mount-util.c:1016]: We don't do that in any other place, e.g. freeze_or_exit_or_reboot() in main. I think if we trust the function in PID1, then we should also here.

Comment (bluca) [reply src/shared/mount-util.c:1016]: fair enough

File: src/basic/fd-util.c [hunk 2]
```diff
@@ -208,10 +208,9 @@ static int get_max_fd(void) {
         return (int) (m - 1);
 }
 
-int close_all_fds(const int except[], size_t n_except) {
+int close_all_fds(int except[], size_t n_except) {
```
Comment (poettering) [inline src/basic/fd-util.c]: hmpf. I really dislike this. we should make parameters constant. mucking around in input params is ugly as fuck.

at the absolute minimum this needs some explicit comment, because it breaks out usual style of having input params as constant, and not meddling with them on output

why not split this function in two: close_all_fds_no_malloc() that doesn't do the close_range() thing, and hence needs no sorting, and close_all_fds() as a wrapper around that which does. Then, from the signal handler call the former, fro the rest the latter.

File: src/basic/fd-util.c [hunk 3]
```diff
@@ -227,129 +226,104 @@ int close_all_fds(const int except[], size_t n_except) {
...
 
                         if (close_range(3, -1, 0) >= 0)
-                                return 1;
+                                return 0;
 
-                        if (!ERRNO_IS_NOT_SUPPORTED(errno) && !ERRNO_IS_PRIVILEGE(errno))
+                        if (ERRNO_IS_NOT_SUPPORTED(errno) || ERRNO_IS_PRIVILEGE(errno))
+                                have_close_range = false;
+                        else
                                 return -errno;
```
Comment (poettering) [inline src/basic/fd-util.c:234]: sounds like an unnecessary change?

**Comment by poettering** [comment]
this sounds like not thought to the end. if malloc() isn't safe from a signal handler, then opendir() isn't either. Hence the whole fallback cannot work.

File: src/core/dbus-execute.c [hunk 4]
```diff
@@ -94,6 +95,10 @@ static int property_get_environment_files(
         return sd_bus_message_close_container(reply);
 }
 
+static bool oom_score_adjust_is_valid(int oa) {
+        return oa >= OOM_SCORE_ADJ_MIN && oa <= OOM_SCORE_ADJ_MAX;
+}
```
Comment (poettering) [inline src/core/dbus-execute.c:100]: hmm, this looks the wrong way round. i think parse_oom_score_adjust(),  should just use this helper too

File: src/basic/process-util.c [hunk 5]
```diff
@@ -1037,30 +1037,6 @@ bool is_main_thread(void) {
         return cached > 0;
 }
 
-_noreturn_ void freeze(void) {
```
Comment (poettering) [inline src/basic/process-util.c:1040]: i am not convinced the reason to move this is sound. we use this from library code, in mount-util.c. Yes, that one is also in src/shared/,  so you get away with moving it, but still: it's fine to call this from some forked off child that some library maintains.

(but i don't really care)

File: src/basic/fd-util.c [hunk 6]
```diff
@@ -208,7 +208,7 @@ static int get_max_fd(void) {
         return (int) (m - 1);
 }
 
-int close_all_fds(int except[], size_t n_except) {
+int close_all_fds_full(int except[], size_t n_except, bool allow_alloc) {
```
Comment (poettering) [inline src/basic/fd-util.c:211]: this is a bit weird. this should be an inner function, and a wrapper, as mentioned
